kmk/modules/steno.py

```diff
@@ -0,0 +1,101 @@
+from kmk.modules import Module
+from kmk.keys import make_key
+
+import usb_cdc
+
+# key order from https://github.com/openstenoproject/plover/blob/main/plover/machine/geminipr.py
+# do not rearrange
+STENO_KEYS = (
+    'STN_FN',
+    'STN_N1',
+    'STN_N2',
+    'STN_N3',
+    'STN_N4',
+    'STN_N5',
+    'STN_N6',
+    'STN_LS1',
+    'STN_LS2',
+    'STN_LT',
+    'STN_LK',
+    'STN_LP',
+    'STN_LW',
+    'STN_LH',
+    'STN_LR',
+    'STN_A',
+    'STN_O',
+    'STN_AS1',
+    'STN_AS2',
+    'STN_RES1',
+    'STN_RES2',
+    'STN_PWR',
+    'STN_AS3',
+    'STN_AS4',
+    'STN_E',
+    'STN_U',
+    'STN_RF',
+    'STN_RR',
+    'STN_RP',
+    'STN_RB',
+    'STN_RL',
+    'STN_RG',
+    'STN_RT',
+    'STN_RS',
+    'STN_RD',
+    'STN_N7',
+    'STN_N8',
+    'STN_N9',
+    'STN_NA',
+    'STN_NB',
+    'STN_NC',
+    'STN_RZ',
+)
+
+
+class Steno(Module):
+    def __init__(self):
+        self._initialize_buffer()
+
+        self._codes = {}
+        for idx, key in enumerate(STENO_KEYS):
+            newKey = make_key(
+                names=(key,), on_press=self._steno_press, on_release=self._steno_release
+            )
+            self._codes[newKey.code] = idx
+
+    def _initialize_buffer(self):
+        self._buffer = bytearray(6)
+        self._buffer[0] = 0x80
+
+    # flip a key's bit in the buffer
+    def _steno_press(self, key, *_):
+        idx = self._codes[key.code]
+        self._buffer[idx // 7] |= 0x80 >> (idx % 7 + 1)
+
+    # send all keys that were pressed, and reset the buffer
+    def _steno_release(self, *_):
+        usb_cdc.data.write(self._buffer)
+        self._initialize_buffer()
```

Comment:
There are two memory consumption optimization that are fairly obvious:
* reuse `self._buffer` instead of creating a new instance
* generate buffer index and code only once during key creation and encode them in `key.code`

That would look like this for example:
```python
class Steno(Module):
    def __init__(self):
        self._buffer = bytearray(6)
        self._initialize_buffer()

        for idx, key in enumerate(STENO_KEYS):
            make_key(
                code=((idx // 7) << 8) | (0x40 >> (idx % 7))
                names=(key,),
                on_press=self._steno_press,
                on_release=self._steno_release
            )

    def _initialize_buffer(self):
        self._buffer[:] = b'\x08\x00\x00\x00\x00\x00'

    # flip a key's bit in the buffer
    def _steno_press(self, key, *_):
        self._buffer[key.code>>8] = key.code & 0xff

    # send all keys that were pressed, and reset the buffer
    def _steno_release(self, *_):
        usb_cdc.data.write(self._buffer)
        self._initialize_buffer()
```

It also looks like you're sending the buffer for every key release, but only the first release sends any meaningful data. The other releases send what I assume is interpreted as empty. This could be further optimized by checking if the buffer contents are non-trivial (or using a flag) to cut down on superflous transmissions. I did not look into how the protocol works.

Reply:
Good points!

I've included your suggestions, and tested that everything is still working as expected.